lib/grandpa/tracker_votes.go

```diff
@@ -0,0 +1,179 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"container/list"
+
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/crypto/ed25519"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+// votesTracker tracks vote messages that could
+// not be processed, and removes the oldest ones once
+// its maximum capacity is reached.
+// It is NOT THREAD SAFE to use.
+type votesTracker struct {
+	// map of vote block hash to authority ID (ed25519 public Key)
+	// to data (peer id + message + tracking linked list element pointer)
+	mapping map[common.Hash]authorityIDToData
+	// double linked list of block hash + authority ID
+	// to track the order vote messages were added in.
+	linkedList *list.List
+	capacity   int
+}
+
+type authorityIDToData map[ed25519.PublicKeyBytes]voteMessageMapData
+
+type voteMessageMapData struct {
+	peerID  peer.ID
+	message *VoteMessage
+	// element contains a blockHashAuthID value which
+	// itself contains a block hash and an authority ID.
+	element *list.Element
+}
+
+type blockHashAuthID struct {
+	blockHash   common.Hash
+	authorityID ed25519.PublicKeyBytes
+}
+
+// newVotesTracker creates a new vote message tracker
+// with the capacity specified.
+func newVotesTracker(capacity int) votesTracker {
+	return votesTracker{
+		mapping:    make(map[common.Hash]authorityIDToData, capacity),
+		linkedList: list.New(),
+		capacity:   capacity,
+	}
+}
+
+func newBlockHashAuthID(blockHash common.Hash,
+	authorityID ed25519.PublicKeyBytes) blockHashAuthID {
+	return blockHashAuthID{
+		blockHash:   blockHash,
+		authorityID: authorityID,
+	}
+}
+
+// add adds a vote message to the vote message tracker.
+// If the vote message tracker capacity is reached,
+// the oldest vote message is removed.
+func (vt *votesTracker) add(peerID peer.ID, voteMessage *VoteMessage) {
+	signedMessage := voteMessage.Message
+	blockHash := signedMessage.Hash
```

Comment:
The hashed signedMessage is the blockHash?

Reply:
It's not the hash of the signed message, but the block hash. Comment on signed message:

```go
// SignedMessage represents a block hash and number signed by an authority
```

I opened #2491 to rename the field to `BlockHash` to avoid any future confusion as well :wink: